src/scaffoldmaker/meshtypes/meshtype_3d_stellate1.py

```diff
@@ -0,0 +1,543 @@
+    :param numArm: number of arms in body
+    :param na: nth arm currently being created
+    :return: x: positions of nodes in arm
+    :return: nodeIdentifier: number of last node in arm +1
+    :return: xnodes_ds1: ds1 derivatives of nodes associated with x
+    :return: xnodes_ds2: ds2 derivatives of nodes associated with y
+    """
+
+    elementsCount1, elementsCount2, elementsCount3 = elementsCount
+    [elen, ewid, zheight] = elementLengths
+
+    xnodes_ds1 = []
+    xnodes_ds2 = []
+    dx_ds1 = [elen, 0.0, 0.0]
+    dx_ds2 = [0.0, ewid, 0.0]
+    wheelDvMult = [0.5, 0.75]
+
+    nodes_per_layer = (elementsCount1 + 1) * (elementsCount2 + 1) - 2 # discount 2 armEnd corners
+    x = []
+
+    # nCentre and rmVertexNodes are in pythonic indexing
+    nCentre = elementsCount1
+    nCentre = [nCentre, nCentre + nodes_per_layer]
+    if na == 0:
+        rmVertexNodes = []
+    elif na != numArm - 1:
+        rmVertexNodes = nCentre + [0, nodes_per_layer]
+    else:
+        rmVertexNodes = nCentre + [0, nodes_per_layer,
+                          2* (elementsCount1+1) - 1, 2* (elementsCount1+1) - 1 + nodes_per_layer ]
+    dcent = [elen * math.cos(armAngleConst / 2), elen * math.sin(armAngleConst / 2), 0.0]
+    dvertex = [round(elen * dipMultiplier * math.cos(thAr), 12),
+                round(elen * dipMultiplier * math.sin(thAr), 12)]
+
+    nid = 0
+    for n3 in range(elementsCount3 + 1):
+        x3 = n3 * zheight
+        for n2 in range(elementsCount2 + 1):
+            for n1 in range(elementsCount1 + 1):
+                # ignore if armEnd corner nodes
+                if n1 == elementsCount1 and (n2 == 0 or n2 == elementsCount2):
+                    pass
+                else:
+                    if n1 == 0 and n2 == 0:
+                        x1 = dvertex[0]
+                        x2 = -dvertex[1]
+                    elif n1 == 0 and n2 == elementsCount2:
+                        x1 = dvertex[0]
+                        x2 = dvertex[1]
+                    else:
+                        x1 = (elen*n1)
+                        x2 = ((n2 - 1) * ewid )
+                    x.append([x1, x2, x3])
+
+                    # DERIVATIVES
+                    ds1 = dx_ds1.copy()
+                    ds2 = dx_ds2.copy()
+                    if n2 == 1 and n1 == 0:
+                        ds2 = dcent
+                        ds1 = [dcent[0], -dcent[1], dcent[2]]
+                    elif n1 == 0 and (n2 == 0 or n2 == elementsCount2):
+                        [p, q] = [x1/dipMultiplier, x2/dipMultiplier] # unit vector
+                        ds2mag = round(2*magnitude([p,q]),6) - magnitude(dcent)
+                        ds2 = [ds2mag*p, ds2mag*q, 0]
+                        ds1 = rotateAboutZAxis(ds2, -math.pi / 2)
+                        for i in range(2):
+                            ds1[i] *= wheelDvMult[0]
+                        ds1 = [d * elen for d in ds1]
```

Comment:
d2 code:
```
ds2 = [dcent[0], -dcent[1], dcent[2]] if (n2 == 0) else dcent
ds2 = setMagnitude(ds2, dipMag)  # must import function from utils.vector
```

Reply:
I am unsure which lines should be replaced by your d2 code: it doesn't seem to work. And did you mean dipMag = ds2mag?

Reply:
I set dipMag in a comment earlier...
The above two lines of mine entirely calculate d2, but you defined d1 from d2 (or at least rotate it 90 degrees). As mentioned in a further comment, it's fine to get the direction of d1 at the dip here, but use the smooth function to get its magnitude later - don't use arbitrary scalings.